docs/topics/competitive-programming.md

```diff
@@ -70,36 +70,80 @@ hash-based maps and sets (`HashMap`/`HashSet`), tree-based ordered maps and sets
 Using a hash-set of integers to track values that were already reached while applying function `f`, 
 the straightforward imperative version of a solution to the problem can be written as shown below:
 
+<tabs>
+<tab title="Kotlin 1.6.0 and later">
+
 ```kotlin
 fun main() {
     var n = readln().toInt() // read integer from the input
     val reached = HashSet<Int>() // a mutable hash set 
     while (reached.add(n)) n = f(n) // iterate function f
     println(reached.size) // print answer to the output
 }
-```
 
-> The readln() function is available since [Kotlin 1.6.0](whatsnew16.md#new-readline-functions).
->
-{type="note"}
+```
 
 There is no need to handle the case of misformatted input in competitive programming. An input format is always precisely
 specified in competitive programming, and the actual input cannot deviate from the input specification in the problem
 statement. That's why you can use Kotlin's [`readln()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/readln.html) function. It asserts that the input string is present and throws
 an exception otherwise. Likewise, the [`String.toInt()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-int.html)
 function throws an exception if the input string is not an integer.
 
+</tab>
+<tab title="Earlier versions">
+
+```kotlin
+fun main() {
+    var n = readLine()!!.toInt() // read integer from the input
+    val reached = HashSet<Int>() // a mutable hash set 
+    while (reached.add(n)) n = f(n) // iterate function f
+    println(reached.size) // print answer to the output
+}
+```
+
+Note the use of Kotlin's
+[null-assertion operator](null-safety.md#the-operator) `!!`
+after the [readLine()](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/read-line.html) function call.
+Kotlin's `readLine()` function is defined to return a
+[nullable type](null-safety.md#nullable-types-and-non-null-types)
+`String?` and returns `null` on the end of the input, which explicitly forces the developer to handle the
+case of missing input.
+
+There is no need to handle the case of misformatted input in competitive programming.
+In competitive programming, an input format is always precisely specified and the actual input cannot deviate from
+the input specification in the problem statement. That's what the null-assertion operator `!!` essentially does &mdash;
```

Comment:
`&mdash;` — just use the mdash in the text, no need to escape it.

IIRC, we use ndash for such things